trackintel/preprocessing/positionfixes.py

```diff
@@ -261,26 +260,30 @@ def generate_triplegs(positionfixes_in, spts=None, method='between_staypoints',
 
         # condition 3: staypoints
         # By our definition the first positionfix after a staypoint is the first positionfix of a tripleg.
-        stp_id = (pfs['staypoint_id'] + 1).fillna(0)
-        cond_stp = (stp_id - stp_id.shift(1)) != 0
+        _stp_id = (pfs['staypoint_id'] + 1).fillna(0)  # this works only for numeric staypoint ids
+        cond_stp = (_stp_id - _stp_id.shift(1)) != 0
+        pfs['cond_stp'] = cond_stp
 
-        # todo: add a "True" to cond_stp at the position that is given in the values of insert_position_spts
+        # special check for case 2: positionfixes that belong to staypoints might not be present in the data.
+        if case == 2:
+            cond_stp = cond_stp | cond_staypoints_case2
 
-        # combine conditions and assign an incrementing id to all positionfixes that start a tripleg
+        # combine conditions
         cond_all = cond_new_user | cond_gap | cond_stp
+        # make sure not to create triplegs within staypoints:
+        cond_all = cond_all & pd.isna(pfs['staypoint_id'])
 
-        # special check for case 2: positionfixes that belong to staypoints might not be present in the data.
-        if case == 2:
-            cond_all = cond_all | cond_staypoints_case2
-            
-        cond_all[~pd.isna(pfs['staypoint_id'])]=False
-            
+        pfs['cond_new_user'] = cond_new_user
+        pfs['cond_gap'] = cond_gap
+        pfs['cond_stp'] = cond_stp
+
+        # todo: A test where you have triplegs with length 1. But most of the cases are solved by
+        #  `cond_all = cond_all & pd.isna(pfs['staypoint_id'])`
         tpls_starts = np.where(cond_all)[0]
         tpls_lengths = np.diff(tpls_starts)
-        cond_to_remove = np.take(tpls_starts,np.where(tpls_lengths<3)[0])
-        for i in cond_to_remove:
-            cond_all[i] = False
-            pfs.loc[i,'tripleg_id'] = -1
+        cond_to_remove = np.take(tpls_starts, np.where(tpls_lengths < 2)[0])  # a valid linestring needs 2 points
+        cond_all.iloc[cond_to_remove] = False
+        pfs.iloc[cond_to_remove].loc['tripleg_id'] = -1
```

Comment:
This causes a warning when I execute it, as it is trying to set the -1 on a copy of pfs

Reply:
I changed this now to `pfs.loc[pfs.index.isin(cond_to_remove), "tripleg_id"] = -1`